src/main/java/io/netty/incubator/codec/http3/QpackDecoder.java

```diff
@@ -16,94 +16,383 @@
+            QpackException.newStatic(QpackDecoder.class, "setDynamicTableCapacity(...)",
+                    "QPACK - decoder dynamic table capacity exceeds max capacity.");
     private static final QpackException HEADER_ILLEGAL_INDEX_VALUE =
             QpackException.newStatic(QpackDecoder.class, "getIndexedHeader(...)", "QPACK - illegal index value");
     private static final QpackException NAME_ILLEGAL_INDEX_VALUE =
-            QpackException.newStatic(QpackDecoder.class, "getIndexedName(...)", "QPACK - illegal index value");
+            QpackException.newStatic(QpackDecoder.class, "getIndexedName(...)", "QPACK - illegal name index value");
+    private static final QpackException INVALID_REQUIRED_INSERT_COUNT =
+            QpackException.newStatic(QpackDecoder.class, "decodeRequiredInsertCount(...)",
+                    "QPACK - invalid required insert count");
+    private static final QpackException MAX_BLOCKED_STREAMS_EXCEEDED =
+            QpackException.newStatic(QpackDecoder.class, "decodeRequiredInsertCount(...)",
+                    "QPACK - exceeded max blocked streams");
+    private static final QpackException BLOCKED_STREAM_RESUMPTION_FAILED =
+            QpackException.newStatic(QpackDecoder.class, "sendInsertCountIncrementIfRequired(...)",
+                    "QPACK - failed to resume a blocked stream");
 
     private static final QpackException UNKNOWN_TYPE =
             QpackException.newStatic(QpackDecoder.class, "decode(...)", "QPACK - unknown type");
 
-    private final QpackHuffmanDecoder huffmanDecoder = new QpackHuffmanDecoder();
+    private final QpackHuffmanDecoder huffmanDecoder;
+    private final QpackDecoderDynamicTable dynamicTable;
+    private final long maxTableCapacity;
+    private final int maxBlockedStreams;
+    private final QpackDecoderStateSyncStrategy stateSyncStrategy;
+    /**
+     * Hashmap with key as the required insert count to unblock the stream and the value a {@link List} of
+     * {@link Runnable} to invoke when the stream can be unblocked.
+     */
+    private final IntObjectHashMap<List<Runnable>> blockedStreams;
+    private long maxEntries;
+    private long fullRange;
+
+    private int blockedStreamsCount;
+    private long lastAckInsertCount;
+
+    QpackDecoder(Http3SettingsFrame localSettings) {
+        this(localSettings, new QpackDecoderDynamicTable(), DEFAULT_SYNC_STRATEGY);
+    }
+
+    QpackDecoder(Http3SettingsFrame localSettings,
+                 QpackDecoderDynamicTable dynamicTable, QpackDecoderStateSyncStrategy stateSyncStrategy) {
+        huffmanDecoder = new QpackHuffmanDecoder();
+        maxTableCapacity = localSettings.getOrDefault(HTTP3_SETTINGS_QPACK_MAX_TABLE_CAPACITY, 0);
+        maxBlockedStreams = toIntExact(localSettings.getOrDefault(HTTP3_SETTINGS_QPACK_BLOCKED_STREAMS, 0));
+        this.stateSyncStrategy = stateSyncStrategy;
+        blockedStreams = new IntObjectHashMap<>(maxBlockedStreams);
```

Comment:
nit: I guess we should have an upper-bound here as otherwise it may be easy to cause an OOME if for example `Integer.MAX_VALUE` or something like this is used.